### DISCUSSION TITLE
General questions about models with dead-time and multiple-input

### DISCUSSION THREAD

**Discussion by TourajFarahmand-AI** [do-mpc/do-mpc#243] [Q&A] [answered]
Hello do-mpc team!
I just started learning about the library and I should say, I'm very impressed. Great work!
I have a couple of general questions about the library which I guess would be helpful for many others too.

1-How would you set up the model/controller/simulator if you have multi-inputs (say two inputs) where one of them is control input (manipulated input) and the other one just acts as an external disturbance to the process (fairly common in chemical processes).
2- How should we deal with models with long dead-time which is pretty common in chemical processing? As I understand there is no explicit way to add it in do-mpc model object.

Many thanks again,
Touraj.

**Comment by 4flixt** [comment, marked as answer]
Hi @TourajFarahmand-AI,

thank you for the compliments :)

1 - Quite easy: We have the variable type (set with [set_variable](https://www.do-mpc.com/en/latest/api/do_mpc.model.Model.set_variable.html#do_mpc.model.Model.set_variable)) ``tvp`` (time-varying parameter). This variable type can be used very flexibly: For future set-points, for uncontrolled exogenous inputs (e.g. ambient temperature) or disturbances. It is important to note that the user must provide a value for this variable for each step of the horizon at each sampling time. Read more about them [here](https://www.do-mpc.com/en/latest/FAQ.html#time-varying-parameters).

2- We suggest to approximate dead-time as a a PT-N system, i.e. by introducing N new states with trivial dynamics. See for example 2.1.3 [here](https://beckassets.blob.core.windows.net/product/readingsample/200291/9781846288289_excerpt_001.pdf) or use a [Pade approximation](https://www.informit.com/articles/article.aspx?p=32090&seqNum=11).

Best,
4flixt

**Reply by anthonycerqui-riotinto** [reply]
Hi @4flixt,
Example 2.1.3 uses Laplace domain - I'd find it really valuable if there was some examples on converting to time domain and setting up a time dependent expression.

Working through 2.1.3 as per your recommendation - they propose to model dead time with:
$G(s) = \frac{1}{(1+KS)^n}$
where $K$ is a time constant and $n$ is an arbitrary(?) integer.

Converting to time domain (I used wolfram), we get:
$y(t)=\frac{ne^{-t/K}(t/K)^{n-1}}{kn!}$

Given this is still a function of time - It's not clear to me how to map this into the required symbolic `_p`, `_u`, `_tvp`, `_x` parameters, and the rhs/other expressions.

Thanks in advance for your help.

**Comment by TourajFarahmand-AI** [comment]
Many thanks!
Both of your suggestions seem to work. Especially second-order Pade approximation since the response time of the process I'm testing is much bigger than the delay.
I have one more question about the rate-of-change constraint which I have in my control problem. Is there an easy way to apply in the API?

Touraj

**Comment by 4flixt** [comment]
Hi @TourajFarahmand-AI ,

rate change constraints can be implemented with a bit of manual tweaking. We suggest to introduce a dummy state that represents the true physical input to the system and a new input which is the rate of change of this dummy state. In this way, you can set hard or soft constraints on the new state (the physical input) and hard constraints on the do-mpc input (which now represents the rate of change.

Best
4flixt

**Comment by TourajFarahmand-AI** [comment]
Hi @4flixt 
A clever way of handling rate-of-change! Thanks,
I also have another general comment that might be helpful for other people who are new to do-mpc. I think the modularity/isolation of do-mpc is quite nice and powerful and at the same time more involved which might be confusing for beginners to the library especially when you use jupyter notebook where all the code is in one file. 
To give you an example, external measured inputs as tvp need to be provided to tvp functions of both controller and simulator since they are separate modules that communicate through input and output. This simple fact was not apparent to me until I spend some time debugging my code :)

Thanks again!
Touraj.

**Reply by 4flixt** [reply]
thank you for the feedback :) 
We always strive to make the documentation of do-mpc even better!

**Reply by TourajFarahmand-AI** [reply]
Hello @4flixt ,
I was watching the webinar on the website. First, I should say it was very informative and thanks for doing that.
At the end of the first notebook, you added a cool slider to see the prediction horizon at simulation time steps. Are there any tricks about the slider widgets and interact command you used in order to update the plot above the cell you created the slider object. Slider and interact command seems to work for me but I need to add a plot.show() in the loop that seems to make it slower and I have to add "%matplotlib widget" to make that work.
Do you think of any workaround? Any changes in new releases of do-mpc wrt. to graphics?

Thanks,
Touraj.
